### PR TITLE
Add analytic solution test for thermoelasticity

### PR DESCRIPTION
**Pull request by stvdwtt**
Add a test using the analytic Eshelby solution and make some minor fixes to MechanicalOperator and MechanicalPhysics.

### CODE REVIEW

File: source/MechanicalOperator.cc [hunk 1]
```diff
@@ -167,51 +167,51 @@ void MechanicalOperator<dim, MemorySpaceType>::assemble_system()
...
-  //         .delta_IJ()
-  //         .dV() -
-  //     dealiiWeakForms::WeakForms::linear_form(test_val,
-  //     rhs_coeff.value(rhs)).dV();
-  assembler +=
-      dealiiWeakForms::WeakForms::bilinear_form(test_grad, stiffness_tensor,
-                                                trial_grad)
-          .dV() -
-      dealiiWeakForms::WeakForms::linear_form(test_val, rhs_coeff.value(rhs))
-          .dV();
```
Comment (Rombur) [inline source/MechanicalOperator.cc:186]: I am guessing you moved this because the test was failing but I would prefer to keep the code structure as it is currently. Enabling thermal expansion is just adding a term on the right-hand-side and the current code makes this clearer. The issue you had is because I used an arbitrary body force in the rhs for testing purpose. I am working on a PR that replaces the arbitrary body force with gravity but if you just want to turn off the body force for now that's fine. 
In general, should we always have the gravity, never have it, or should we allow to turn it on/off (maybe only for testing)? I was planning to use gravity as the body force because:
 1. what we are doing right now is dumb
 2. I have no idea when it becomes important to take gravity into account

File: source/MechanicalPhysics.cc [hunk 2]
```diff
@@ -68,8 +68,8 @@ void MechanicalPhysics<dim, MemorySpaceType>::setup_dofs(
     dealii::LA::distributed::Vector<double, dealii::MemorySpace::Host> const
         &temperature)
 {
-  setup_dofs();
   _mechanical_operator->update_temperature(thermal_dof_handler, temperature);
+  setup_dofs();
```
Comment (Rombur) [inline source/MechanicalPhysics.cc:72]: I am fine with the change but does it really matters? I don't see how it changes the computation.

File: tests/test_mechanical_physics.cc [hunk 3]
```diff
@@ -240,3 +243,169 @@ BOOST_AUTO_TEST_CASE(elastostatic)
   for (unsigned int i = 0; i < reference_solution.size(); ++i)
     BOOST_TEST(solution[i] == reference_solution[i], tt::tolerance(tolerance));
 }
+
+template <int dim>
+class InitialValueT : public dealii::Function<dim>
+{
+public:
+  virtual double value(const dealii::Point<dim> &p,
```
Comment (Rombur) [inline tests/test_mechanical_physics.cc:251]: You don't need `virtual` if you have `override`. It is done like that in deal.II because someone thought it was clearer for C++ beginner that the function was still `virtual`

File: tests/test_mechanical_physics.cc [hunk 4]
```diff
@@ -240,3 +243,169 @@ BOOST_AUTO_TEST_CASE(elastostatic)
...
+  }
+};
+
+namespace utf = boost::unit_test;
+
+/*
+ * This unit test uses the analytic solution for the displacement around a
+ * spherical inclusion as a test case. As a result of the boundary conditions
+ * (fixed on one face) and the need to keep the test computationally
+ * inexpensive, a very loose tolerance has been chosen. Also for simplicity, the
```
Comment (Rombur) [inline tests/test_mechanical_physics.cc:275]: That's fine but have you tried on a large domain to check that we get the correct results? At some point, I want to add nightly tests and this test would be a good candidate for that. So please add in the comments the parameters which give an accurate result.

File: source/MechanicalOperator.cc (hunk 1 above)
Comment (stvdwtt) [reply source/MechanicalOperator.cc:186]: You're right, that was my reasoning.

As far as gravity goes -- for most cases we're going to deal with gravity will be negligible. I can see a few cases where gravity might matter (polymers, very large cantilevered parts), but nothing in the near term. 

I recommend adding an option to turn it on/off in the constructor (default off). Having the ability to add a body force could be useful for testing and it's not much extra code to maintain.

If having it optionally on works for you, I can modify the code in this PR to do that (but leaving the body force as the current arbitrary one).

Comment (Rombur) [reply source/MechanicalOperator.cc:186]: sounds good

File: source/MechanicalPhysics.cc (hunk 2 above)
Comment (stvdwtt) [reply source/MechanicalPhysics.cc:72]: The `MechanicalOperator::reinit` call needs `MechanicalOperator::_thermal_dof_handler` to be set.

File: source/MechanicalOperator.cc (hunk 1 above)
Comment (stvdwtt) [reply source/MechanicalOperator.cc:186]: I've added the optional addition of the body force.

File: tests/test_mechanical_physics.cc (hunk 3 above)
Comment (stvdwtt) [reply tests/test_mechanical_physics.cc]: Fixed. (And thanks for explaining why that's the standard way it's done in deal.ii)

File: tests/test_mechanical_physics.cc (hunk 4 above)
Comment (stvdwtt) [reply tests/test_mechanical_physics.cc]: I tweaked the base test to have an adaptive mesh so it gets to a reasonable solution faster. I also included a comment about how to get to less than 5% error.

File: source/MechanicalOperator.cc [hunk 5]
```diff
@@ -167,51 +168,52 @@ void MechanicalOperator<dim, MemorySpaceType>::assemble_system()
...
 
   std::unique_ptr<dealii::hp::FEValues<dim>> temperature_hp_fe_values;
+
   // If the initial temperature is positive, we solve the thermoelastic problem.
   if (_initial_temperature >= 0.)
   {
+    assembler += dealiiWeakForms::WeakForms::bilinear_form(
+                     test_grad, stiffness_tensor, trial_grad)
+                     .dV();
```
Comment (Rombur) [inline source/MechanicalOperator.cc:179]: I should have been clearer. Move the `bilinear_form` before the `if`. So that the `if(_initial_temperature)` and the ` if (_include_gravity)` only include the `linear_form`, i.e., the right-hand-side.

**Review by Rombur — APPROVED** [review]

**Comment by Rombur** [comment]
Looks good. I'll merge once the CI passes
